program_transformers/src/rollups/rollup_persister.rs

```diff
@@ -0,0 +1,616 @@
+    txn: Arc<T>,
+    notification_receiver: Receiver<()>,
+    downloader: D,
+}
+
+pub struct RollupDownloaderForPersister {}
+
+#[async_trait]
+impl RollupDownloader for RollupDownloaderForPersister {
+    async fn download_rollup(&self, url: &str) -> Result<Box<Rollup>, RollupValidationError> {
+        let response = reqwest::get(url).await?.bytes().await?;
+        Ok(Box::new(serde_json::from_slice(&response)?))
+    }
+
+    async fn download_rollup_and_check_checksum(
+        &self,
+        url: &str,
+        checksum: &str,
+    ) -> Result<Box<Rollup>, RollupValidationError> {
+        let response = reqwest::get(url).await?.bytes().await?;
+        let file_hash = xxhash_rust::xxh3::xxh3_128(&response);
+        let hash_hex = hex::encode(file_hash.to_be_bytes());
+        if hash_hex != checksum {
+            return Err(RollupValidationError::InvalidDataHash(
+                checksum.to_string(),
+                hash_hex,
+            ));
+        }
+        Ok(Box::new(serde_json::from_slice(&response)?))
+    }
+}
+
+impl<T: ConnectionTrait + TransactionTrait, D: RollupDownloader> RollupPersister<T, D> {
+    pub fn new(txn: Arc<T>, notification_receiver: Receiver<()>, downloader: D) -> Self {
+        Self {
+            txn,
+            notification_receiver,
+            downloader,
+        }
+    }
+
+    pub async fn persist_rollups(&self) {
+        loop {
+            if let Err(e) = self.notification_receiver.recv().await {
+                error!("Recv rollup notification: {}", e);
+                tokio::time::sleep(Duration::from_secs(5)).await;
+                continue;
+            };
+            let Ok((rollup_to_verify, rollup)) = self.get_rollup_to_verify().await else {
+                statsd_count!("rollup.fail_get_rollup", 1);
+                continue;
+            };
+            let Some(rollup_to_verify) = rollup_to_verify else {
+                // no rollups to persist
+                continue;
+            };
+            let rollup = rollup
+                .map(|r| bincode::deserialize::<Rollup>(r.rollup_binary_bincode.as_slice()))
+                .transpose()
+                .unwrap_or_default();
+            self.persist_rollup(rollup_to_verify, rollup.map(Box::new))
+                .await;
+        }
+    }
+
+    pub async fn persist_rollup(
+        &self,
+        mut rollup_to_verify: rollup_to_verify::Model,
+        mut rollup: Option<Box<Rollup>>,
+    ) {
+        let start_time = Instant::now();
+        info!("Persisting {} rollup", &rollup_to_verify.url);
+        loop {
+            match &rollup_to_verify.rollup_persisting_state {
+                &RollupPersistingState::ReceivedTransaction => {
+                    // We get ReceivedTransaction state on the start of processing
+                    rollup_to_verify.rollup_persisting_state =
+                        RollupPersistingState::StartProcessing;
```

Comment:
This case should not be possible right because anything gotten by `get_rollup_to_verify()` should be moved out of this state before being sent as a parameter to `persist_rollup`?